sdks/java/core/src/main/java/org/apache/beam/sdk/util/WindowingStrategies.java

```diff
@@ -0,0 +1,220 @@
+      case DISCARDING:
+        return AccumulationMode.DISCARDING_FIRED_PANES;
+      case ACCUMULATING:
+        return AccumulationMode.ACCUMULATING_FIRED_PANES;
+      case UNRECOGNIZED:
+      default:
+        // Whether or not it is proto that cannot recognize it (due to the version of the
+        // generated code we link to) or the switch hasn't been updated to handle it,
+        // the situation is the same: we don't know what this OutputTime means
+        throw new IllegalArgumentException(
+            String.format(
+                "Cannot convert unknown %s to %s: %s",
+                RunnerApi.AccumulationMode.class.getCanonicalName(),
+                AccumulationMode.class.getCanonicalName(),
+                proto));
+    }
+  }
+
+  public static RunnerApi.AccumulationMode toProto(AccumulationMode accumulationMode) {
+    switch (accumulationMode) {
+      case DISCARDING_FIRED_PANES:
+        return RunnerApi.AccumulationMode.DISCARDING;
+      case ACCUMULATING_FIRED_PANES:
+        return RunnerApi.AccumulationMode.ACCUMULATING;
+      default:
+        throw new IllegalArgumentException(
+            String.format(
+                "Cannot convert unknown %s to %s: %s",
+                AccumulationMode.class.getCanonicalName(),
+                RunnerApi.AccumulationMode.class.getCanonicalName(),
+                accumulationMode));
+    }
+  }
+
+  public static RunnerApi.ClosingBehavior toProto(Window.ClosingBehavior closingBehavior) {
+    switch (closingBehavior) {
+      case FIRE_ALWAYS:
+        return RunnerApi.ClosingBehavior.EMIT_ALWAYS;
+      case FIRE_IF_NON_EMPTY:
+        return RunnerApi.ClosingBehavior.EMIT_IF_NONEMPTY;
+      default:
+        throw new IllegalArgumentException(
+            String.format(
+                "Cannot convert unknown %s to %s: %s",
+                ClosingBehavior.class.getCanonicalName(),
+                RunnerApi.ClosingBehavior.class.getCanonicalName(),
+                closingBehavior));
+    }
+  }
+
+  public static RunnerApi.OutputTime toProto(OutputTimeFn<?> outputTimeFn) {
+    if (outputTimeFn instanceof WindowingStrategy.CombineWindowFnOutputTimes) {
+      return OutputTimeFns.toProto(
+          ((CombineWindowFnOutputTimes<?>) outputTimeFn).getOutputTimeFn());
+    } else {
+      return OutputTimeFns.toProto(outputTimeFn);
+    }
+  }
+
+  // This URN says that the coder is just a UDF blob the indicated SDK understands
+  private static final String CUSTOM_CODER_URN = "urn:beam:coders:custom:1.0";
+
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+  public static RunnerApi.MessageWithComponents toProto(WindowFn<?, ?> windowFn)
+      throws IOException {
+    Coder<?> windowCoder = windowFn.windowCoder();
+
+    // TODO: re-use components
+    String windowCoderId = UUID.randomUUID().toString();
+    String customCoderId = UUID.randomUUID().toString();
+
+    return RunnerApi.MessageWithComponents.newBuilder()
```

Comment:
Yes, we surely will, with a reverse lookup. I have that planned for a further change, since this one hit ~500 LOC and is probably pretty easy to review, unblocking me to work asynchronously ahead. If you like I could do it here.